### PR TITLE
Add has_secure_token to Active Record

### PR DESCRIPTION
**Pull request by robertomiranda**
I take the audacity of include has_scure_token to Active Model since the implementation is using methods that are implemented in most of popular ORMs or ODMs in ruby, like `exists?`and `save`, is just as suggestion and for sure I'm open to move it to Active Record anyway :smile:. Is still missing documentation but I'd like to hear first opinions about the implementation/code

**Usage**

``` ruby
class User < ActiveRecord::Base
  has_secure_token :token1, :token2, key_length: 30
end

user = User.new
user.save
user.token1 # => "973acd04bc627d6a0e31200b74e2236"
user.token2 # => "e2426a93718d1817a43abbaa8508223"
user.regenerate_token1! # => true
user.regenerate_token2! # => true
```

ref #16879

cc @dhh

### CODE REVIEW

**Comment by hundredwatt** [comment]
`Devise` recently moved from storing tokens directly to storing an HMAC digest of the token: https://github.com/plataformatec/devise/commit/143794d701bcd7b8c900c5bb8a216026c3c68afc

From a security perspective, should that be considered here?

File: activerecord/test/cases/secure_token_test.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  end
+
+  test "raise and exception when with 10 attemps is reached" do
+    User.stubs(:exists?).returns(*Array.new(10, true))
+    assert_raises(RuntimeError) do
+      @user.save
+    end
+  end
+
+  test "assing unique token after 9 attemps reached" do
```
Comment (robertomiranda) [inline activerecord/test/cases/secure_token_test.rb]: > @dhh: We still need another test case for "9 collisions, then a success = win" and "10 collisions exactly = fail" to test that the main loop is functioning as it should.

done!

**Comment by dhh** [comment]
This is looking good to me now :+1:. I say we still want to do the base62 upgrade, but we can do that after merging.

**Comment by robertomiranda** [comment]
@dhh great, once this PR be already merged I'll open another one with base62 upgrade :+1:

**Comment by dhh** [comment]
There you go :)

**Comment by simi** [comment]
:clap:

File: activerecord/lib/active_record/secure_token.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        # Load securerandom only when has_secure_key is used.
+        require 'securerandom'
+        define_method("regenerate_#{attribute}") { update! attribute => self.class.generate_unique_secure_token(attribute) }
+        before_create { self.send("#{attribute}=", self.class.generate_unique_secure_token(attribute)) }
+      end
+
+      def generate_unique_secure_token(attribute)
+        10.times do |i|
+          SecureRandom.hex(12).tap do |token|
+            if exists?(attribute => token)
```
Comment (chancancode) [inline activerecord/lib/active_record/secure_token.rb]: Sorry I missed this earlier... we can probably remove this when @robertomiranda get to switching to base 62. A random UUID (**Universally Unique** Identifier) has 122 random bits. [Wikipedia](http://en.wikipedia.org/wiki/Universally_unique_identifier#Random_UUID_probability_of_duplicates) puts that into perspective:

> To put these numbers into perspective, the annual risk of a given person being hit by a meteorite is estimated to be one chance in 17 billion,[36] which means the probability is about 0.00000000006 (6 × 10−11), equivalent to the odds of creating a few tens of trillions of UUIDs in a year and having one duplicate. In other words, only after generating 1 billion UUIDs every second for the next 100 years, the probability of creating just one duplicate would be about 50%.

With a 24 character base62 sequence, we will have 130 random bits and some change, which makes it over 256 times less likely than a random UUID collision. So basically if a collision is detected, it is much probably more likely that the code or hardware enforcing the uniqueness is at fault than an actual violation, or perhaps a hacker replaced your entropies on your system with anchovies :fish:

Either way, if it happens, you probably have bigger problem to worry about elsewhere :smile:

That said, a uniqueness constraint is still very valuable for protecting against other kind of errors, such as accidentally importing a row from an external source twice, so it should be probably encouraged. What would be nice is that if you can generate this from the generator, i.e. ...

```
rails g model user ... reset_password_token:token
```

...which adds the string column, adds the unique constraint, and inject the `has_secure_token` call in the model. (Same thing for migration generator, scaffold, etc.)

For "plug and play", if you decided not to use a unique constraint, the base62 version should give you the same level of protection... the chance of a collision with the `exists` check *and* a race condition happening at the same time is probably comparable to the chance of collision here (which is, again, practically impossible :smile:)

Comment (chancancode) [inline activerecord/lib/active_record/secure_token.rb]: cc @dhh, what do you think about adding this to the migration generator?

Comment (dhh) [inline activerecord/lib/active_record/secure_token.rb]: Me likey :+1: – and sounds good to drop the retry with base62.

Comment (chancancode) [inline activerecord/lib/active_record/secure_token.rb]: Great :grin: @robertomiranda are you still interested in pursuing these?

Comment (robertomiranda) [inline activerecord/lib/active_record/secure_token.rb]: @chancancode yes, I am. I think that the path here is:

1. Add Base62 monkey patch to ActiveSupport
2. Switch to Base62 for generates secure tokens and remove the retry
3. Add the secure tokens migration helper and generator
4. Move SecureToken to ActiveModel, in favor of "plug and play"

**Comment by jonatack** [comment]
:green_heart:

File: activerecord/test/cases/secure_token_test.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
+require 'cases/helper'
+require 'models/user'
+
+class SecureTokenTest < ActiveRecord::TestCase
+  def setup
+    @user = User.new
+  end
+
+  def test_assing_unique_token_values
```
Comment (simi) [inline activerecord/test/cases/secure_token_test.rb]: Is this preferred way over `def test`?

File: activerecord/lib/active_record/secure_token.rb [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+      #
+      # SecureRandom is used to generate the 24-character unique token, so collisions are highly unlikely.
+      # We'll check to see if the generated token has been used already using #exists?, and retry up to 10
+      # times to find another unused token. After that a RuntimeError is raised if the problem persists.
+      #
+      # Note that it's still possible to generate a race condition in the database in the same way that
+      # validates_presence_of can. You're encouraged to add a unique index in the database to deal with
+      # this even more unlikely scenario.
+      def has_secure_token(attribute = :token)
+        # Load securerandom only when has_secure_key is used.
```
Comment (aripollak) [inline activerecord/lib/active_record/secure_token.rb]: `s/key/token/`?

**Comment by aripollak** [comment]
It seems like @hundredwatt's comment wasn't addressed here. It seems weird to me to call something a secure token if it's stored in the database as plaintext, when the equivalent for passwords, `has_secure_password`, stores its value encrypted.

**Comment by dhh** [comment]
The security is in terms of its random generation. See SecureRandom.

File: activerecord/lib/active_record/secure_token.rb [hunk 5]
```diff
@@ -0,0 +1,49 @@
+module ActiveRecord
+  module SecureToken
+    extend ActiveSupport::Concern
+
+    module ClassMethods
+      # Example using has_secure_token
+      #
+      #   # Schema: User(toke:string, auth_token:string)
```
Comment (abscondite) [inline activerecord/lib/active_record/secure_token.rb]: missed 'n'

Comment (maclover7) [inline activerecord/lib/active_record/secure_token.rb]: Fixed via #18537.

File: activerecord/lib/active_record/secure_token.rb (hunk 2 above)
Comment (chabgood) [inline activerecord/lib/active_record/secure_token.rb]: By "it adds a unique constraint"  is this an index with a unique constraint on that column?

**Pull request merged**